Fluid dynamics automatic time step utility

rubenzorrilla opened the pull request:
From the old time step utility in the incompressible_fluid_application a new automatic time step utility has been created (according to the new format) and added to the Python solvers.

rubenzorrilla committed:
New estimate dt utility

Creation of a new estimate dt utility.

rubenzorrilla committed:
Automatic time stepping advances

Addition of automatic time stepping utility to python solvers. Minor modifications in the EstimateDtUtility.

rubenzorrilla committed:
Python solvers time stepping optimization

rubenzorrilla committed:
Update navier_stokes_base_solver.py

jcotela commented on applications/FluidDynamicsApplication/custom_utilities/estimate_dt_utilities.h:
```diff
@@ -0,0 +1,275 @@
+                    Proj = fabs(Proj);
+                    if (Proj > ElemProj) ElemProj = Proj;
+                }
+                itElem->SetValue(DIVPROJ,ElemProj*Dt);
```
This recycled variable is not ideal here. Do we have a better one? Should we create a new one?

RiccardoRossi replied:
do we actually need it? can't we do without?

rubenzorrilla replied:
I completely agree. We can create a CFLNUMBER variable in the FluidDynamicsApplication despite it will be rarely used.

jcotela replied:
I think it is good if we create it. It is not a problem if it is rarely used, since unused variables are 'free'. Might I suggest CFL_NUMBER with an underscore as the name?

pooyan-dadvand replied:
👍 for Jordi: CFL_NUMBER

jcotela commented on applications/FluidDynamicsApplication/python_scripts/navier_stokes_embedded_solver.py:
```diff
@@ -105,7 +110,7 @@ def AddVariables(self):
         super(NavierStokesEmbeddedMonolithicSolver, self).AddVariables()
         ## Add specific variables needed for the embedded solver
-        self.main_model_part.AddNodalSolutionStepVariable(KratosMultiphysics.NODAL_H)           # Nodal volume necessary for refinement
+        # self.main_model_part.AddNodalSolutionStepVariable(KratosMultiphysics.NODAL_H)           # Nodal volume necessary for refinement
```
Do not comment out unused lines, remove them (if you ever need them again, they will be somewhere in the git history)

jcotela commented on applications/FluidDynamicsApplication/python_scripts/navier_stokes_solver_vmsmonolithic.py:
```diff
@@ -58,12 +58,16 @@ def __init__(self, main_model_part, custom_settings):
             "MoveMeshFlag": false,
-            "use_slip_conditions": false,
             "turbulence_model": "None",
             "use_spalart_allmaras": false
```
If we remove the slip_conditions, the turbulence_model and spalart_allmaras flags should go too (they have a very limited support at the moment, I would hide them from the interface). The same might apply for the regularization_coef.

RiccardoRossi commented on applications/FluidDynamicsApplication/custom_python/add_custom_utilities_to_python.cpp:
```diff
@@ -1,22 +1,22 @@
 /// KratosFluidDynamicsApplication
```
use the simplifies header

RiccardoRossi commented on applications/FluidDynamicsApplication/custom_utilities/estimate_dt_utilities.h:
```diff
@@ -0,0 +1,275 @@
+     * nodal and elemental values of the variable in GiD)
+     * @param Dt The time step used by the fluid solver
+     */
+    void CalculateLocalCFL(const double Dt)
```
@rubenzorrilla WHY do we need to store the local CFL number?

jcotela replied:
Is was stored for visualization/debug. I think it is a good thing.

pooyan-dadvand commented on applications/FluidDynamicsApplication/custom_utilities/estimate_dt_utilities.h:
```diff
@@ -0,0 +1,275 @@
+
+    double              mCFL; // A reference to the user-defined CFL number
+    double              mDtMax; // A reference to the user-defined maximum time increment allowed
+    ModelPart           mrModelPart; // A reference to the problem's model part
```
mrModelPart should be a reference but the "&" is missing. isn't it?

jcotela replied:
:+1: But, since you mention it, maybe storing a ModelPart::Pointer would be safer? (since the reference comes from python and might expire anytime)

pooyan-dadvand commented on applications/FluidDynamicsApplication/custom_utilities/estimate_dt_utilities.h:
```diff
@@ -0,0 +1,275 @@
+    ///@name Serialization
+    ///@{
+
+    //friend class Serializer;
```
I would suggest to take out all of the section comments and commented parts which you are not using in order to reduce the file size and the amount of scroll one should do.

pooyan-dadvand replied:
👍 for Riccardo

jcotela committed:
Cleanup of estimate_dt_utilities.h

rubenzorrilla committed:
Changing old fashioned header

rubenzorrilla committed:
Fluid dynamics Python solvers clean up

jcotela committed:
Removing reduction(max) pragma

Eliminating a call to #pragma omp reduce(max:var), which is part of OpenMP 3.1 and not supported by Visual Studio 2015

rubenzorrilla committed:
Compilation warnings removal

Comparison between unsigned int and int.

jcotela commented:
@rubenzorrilla can you update navier_stokes_embedded_solver.py? I'll approve my review once this is done.

rubenzorrilla committed:
Bug fix in automatic time stepping

A minimum time step is now set to avoid division by zero.

rubenzorrilla committed:
Python solvers update and clean up

Clean up and automatic time step addition in Trilinos fluid dynamics solvers. Modifications according to the minimum time step buf fix.

rubenzorrilla committed:
Python solvers update and clean up

Clean up and automatic time step addition in Trilinos fluid dynamics solvers. Modifications according to the minimum time step buf fix.

rubenzorrilla committed:
Merge trilinos fractional step solver

rubenzorrilla committed:
Trilinos navier stokes embedded solver automatic time stepping

rubenzorrilla commented:
@jcotela Python solvers amended (both serial and Trilinos ones). Btw, I also did the parameters clean up that you suggested.

jcotela reviewed (changes requested):
Careful, this looks like a conflict

jcotela committed:
Removing conflicted line

pooyan-dadvand reviewed (commented):
👍

rubenzorrilla committed:
Remove unused commented lines

jcotela commented:
@pooyan-dadvand Can you please re-review this pull request?

pooyan-dadvand commented on applications/FluidDynamicsApplication/custom_utilities/estimate_dt_utilities.h:
```diff
@@ -221,8 +249,9 @@ class EstimateDtUtility
+    double              mCFL;        // User-defined CFL number
+    double              mDtMax;      // User-defined maximum time increment allowed
+    double              mDtMin;      // User-defined minimum time increment allowed
     ModelPart::Pointer  mpModelPart; // The problem's model part
```
Using shared_ptr (ModelPart::Pointer) is ok for python but gives a serious problem in cpp. The reason is the fact that submodelparts are stored by raw pointers and there is no way to get a Pointer from them. So I would suggest to use a normal pointer. Also it is arguable (and we can discuss it as another issue) if I should keep the sub modelparts by Pointer.

jcotela committed:
Working with a model part reference (not a counted pointer)

roigcarlo commented:
@pooyan-dadvand Please ignore the appveyor checks, I was setting up the windows builds and some commits were wrongly targeted. It should be resolved now

pooyan-dadvand reviewed (approved):
👍

pooyan-dadvand commented:
The test is experimental and failing due to the configuration. So you may merge it without problem.
